src/blockchain/backend.rs

```diff
@@ -0,0 +1,145 @@
+use std::fs;
+use std::path::Path;
+use std::sync::{Arc, RwLock};
+
+use anyhow::{anyhow, Result};
+use db_key::Key;
+use leveldb::database::Database;
+use leveldb::kv::KV;
+use leveldb::options::{Options, ReadOptions, WriteOptions};
+
+use crate::blockchain::header::{BackendHeader, HeaderMeta, HeaderMetaCache};
+use crate::blockchain::utils::{read_meta, wrap_with_level_db, wrap_with_rocks_db};
+use crate::blockchain::{utils, BackendDatabase, HeaderBackend};
+use crate::core::block::BlockId;
+use crate::core::header::Header;
+use crate::core::U256;
+
+pub struct BlockChainDB {
+    database: Arc<dyn BackendDatabase>,
```

Comment:
Why not use a generic parameter?
`pub struct BlockChainDB<DB: BackendDatabase> {
    database: Arc<DB>,
    ...
}`